hera_cal/smooth_cal.py

```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
             obtained from hera_cal.smooth_cal.dpss_filters
-        XTXinv: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. Useful for filtering
+        method: method to use for solving the linear least squares problem. Options are 'pinv', 'lstsq', 'lu_solve', and 'solve'.
+            'pinv' uses np.linalg.pinv to compute the least squares solution and tends to be the most reliable. 'lu_solve'
+            uses scipy.linalg.lu_solve to compute the least squares solution and tends to be the fastest. 
+            'solve' uses np.linalg.solve and 'lstsq' uses np.linalg.lstsq and have comparable results.
+        cached_input: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. Useful for filtering
             many gain grids with similar flagging patterns. np.ndarray of shape (N_time_vectors * N_freq_vectors,
-            N_time_vectors * N_freq_vectors)
+            N_time_vectors * N_freq_vectors). Can be obtained using the 'cached_output' return value from a
+            previous call to this function.
+        XTXinv: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. 
+            Warning: This is deprecated and will be removed in a future version. Use cached_input instead.
 
     Returns:
         filtered: filtered gain grids from least squares fit, np.ndarray with the
             same shape as the input gain grid
-        info: dictionary containing fit components and XTXinv
+        cached_output: dictionary containing intermediated cached products for linear fitting. Useful
+            for use in subsequent filtering if the input gains have the same flagging pattern previous input gains.
     """
-    # If (X^T W X)^-1 is not precalculated, calculate it
-    if XTXinv is None:
-        # Make sure data types between weights and filters are compatible for einsum optimization
-        weights = weights.astype(time_filters.dtype)
-
-        # Use einsum to calculate (X^T W X) in a memory efficient way
-        XTX = np.einsum(
-            "ij,kl,jl,jm,ln->ikmn", np.transpose(time_filters), np.transpose(freq_filters),
-            weights, time_filters, freq_filters, optimize=True
+    # Check for deprecated XTXinv argument
+    if XTXinv is not None:
+        warnings.warn(
+            "The XTXinv argument is deprecated and will be removed in a future version. Use cached_input instead.",
+            DeprecationWarning,
         )
-        ncomps = time_filters.shape[1] * freq_filters.shape[1]
-        XTX = np.reshape(XTX, (ncomps, ncomps))
-        XTXinv = np.linalg.pinv(XTX)
-
+    # Make sure data types between weights and filters are compatible for einsum optimization
+    weights = weights.astype(time_filters.dtype)
+
+    # Calculate the number of components in the design matrix
+    ncomps = time_filters.shape[1] * freq_filters.shape[1]
+
+    # Use jax einsum to calculate (X^T W X) in a memory efficient way
+    XTX = jnp.einsum(
```

Comment:
If we have the cache, is it possible to avoid this calculation? Would that be worthwhile?